File: ueims_frontend/src/services/RequestLogService.ts
import { api } from './api';

export type HttpMethod = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE' | 'OPTIONS' | 'HEAD';

export interface RequestLogEntry {
  id: string;
  userId?: string;
  userEmail?: string;
  sessionId?: string;
  method: HttpMethod;
  endpoint: string;
  statusCode?: number;
  ipAddress?: string;
  userAgent?: string;
  responseTimeMs?: number;
  timestamp: string;
}

export interface RequestLogParams {
  userId?: string;
  method?: HttpMethod;
  endpoint?: string;
  startDate?: string;
  endDate?: string;
  page?: number;
  size?: number;
}

export const RequestLogService = {
  getLogs: async (params?: RequestLogParams) => {
    const res = await api.get('/request-logs', { params });
    return res.data?.result ?? res.data;
  },

  getLogsByUser: async (userId: string, page = 0, size = 20) => {
    const res = await api.get(`/request-logs/user/${userId}`, { params: { page, size } });
    return res.data?.result ?? res.data;
  },

  getRecentLogs: async (page = 0, size = 20) => {
    const res = await api.get('/request-logs/recent', { params: { page, size } });
    return res.data?.result ?? res.data;
  },

  exportCsv: async (params?: RequestLogParams) => {
    const res = await api.get('/request-logs/export', {
      params,
      responseType: 'blob',
    });
    return res.data;
  },

  clearAll: async () => {
    const res = await api.delete('/request-logs');
    return res.data?.result ?? 0;
  },
};

File: ueims_backend/src/main/java/com/ueims/service/RequestLogService.java
package com.ueims.service;

import java.time.LocalDateTime;
import java.util.UUID;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import com.ueims.dto.response.RequestLogResponseDTO;
import com.ueims.model.entity.RequestLog;
import com.ueims.model.entity.RequestLog.HttpMethod;

public interface RequestLogService {
    void logRequest(RequestLog requestLog);

    Page<RequestLogResponseDTO> searchLogs(
            UUID userId,
            HttpMethod method,
            String endpoint,
            LocalDateTime startDate,
            LocalDateTime endDate,
            Pageable pageable);

    Page<RequestLogResponseDTO> getLogsByUser(UUID userId, Pageable pageable);

    Page<RequestLogResponseDTO> getRecentLogs(Pageable pageable);

    byte[] exportCsv(UUID userId, HttpMethod method, String endpoint, LocalDateTime startDate, LocalDateTime endDate);

    long clearAll();
}
